File: src/geometry/wm/convert/vector.ts
import { extendBBox } from '../../bbox.js';

import type {
  BBOX,
  Feature,
  Geometry,
  MValue,
  Point,
  Point3D,
  Properties,
  VectorCoordinates,
  VectorFeature,
  VectorGeometry,
  VectorPoint,
} from '../../index.js';

/**
 * Convert a GeoJSON Feature to a GeoJSON Vector Feature
 * @param data - GeoJSON Feature
 * @param buildBBox - optional - build a bbox for the feature if desired
 * @returns - GeoJson Vector Feature
 */
export function toVector<
  M = Record<string, unknown>,
  D extends MValue = Properties,
  P extends Properties = Properties,
  G extends Geometry<D> = Geometry<D>,
>(data: Feature<M, D, P, G>, buildBBox?: boolean): VectorFeature<M, D, P, VectorGeometry<D>> {
  const { id, properties, metadata } = data;
  const vectorGeo = geoToVector<D>(data.geometry, buildBBox);
  return {
    id,
    type: 'VectorFeature',
    properties,
    metadata,
    geometry: vectorGeo,
  };
}

/**
 * Convert a GeoJSON Geometry to an Vector Geometry
 * @param geometry - GeoJSON Geometry
 * @param buildBBox - optional - build a bbox for the feature if desired
 * @returns - GeoJson Vector Geometry
 */
export function geoToVector<M extends MValue = Properties>(
  geometry: Geometry<M>,
  buildBBox?: boolean,
): VectorGeometry<M> {
  const { type, coordinates: coords, mValues, bbox } = geometry;
  const newBBox: BBOX | undefined =
    buildBBox !== false && bbox === undefined ? ([] as unknown as BBOX) : undefined;

  let coordinates: VectorCoordinates;
  if (type === 'Point' || type === 'Point3D') coordinates = convertPoint(coords, mValues, newBBox);
  else if (type === 'MultiPoint' || type === 'MultiPoint3D')
    coordinates = coords.map((point, i) => convertPoint(point, mValues?.[i], newBBox));
  else if (type === 'LineString' || type === 'LineString3D')
    coordinates = coords.map((point, i) => convertPoint(point, mValues?.[i], newBBox));
  else if (type === 'MultiLineString' || type === 'MultiLineString3D')
    coordinates = coords.map((line, i) =>
      line.map((point, j) => convertPoint(point, mValues?.[i]?.[j], newBBox)),
    );
  else if (type === 'Polygon' || type === 'Polygon3D')
    coordinates = coords.map((line, i) =>
      line.map((point, j) => convertPoint(point, mValues?.[i]?.[j], newBBox)),
    );
  else if (type === 'MultiPolygon' || type === 'MultiPolygon3D')
    coordinates = coords.map((polygon, i) =>
      polygon.map((line, j) =>
        line.map((point, k) => convertPoint(point, mValues?.[i]?.[j]?.[k], newBBox)),
      ),
    );
  else {
    throw new Error('Invalid GeoJSON type');
  }
  const is3D = type.slice(-2) === '3D';
  // @ts-expect-error - coordinates complains, but the way this is all written is simpler
  return { type: type.replace('3D', ''), is3D, coordinates, bbox: newBBox ?? bbox };
}

/**
 * Mutate a GeoJSON Point to a GeoJSON Vector Point
 * @param point - GeoJSON flat Point
 * @param m - optional m-value
 * @param bbox - if bbox is provided, we will extend the bbox
 * @returns - GeoJSON Vector Point
 */
function convertPoint(point: Point | Point3D, m?: MValue, bbox?: BBOX): VectorPoint {
  const newPoint: VectorPoint = { x: point[0], y: point[1], z: point[2], m };
  if (bbox !== undefined) {
    const newBBox = extendBBox(bbox, newPoint);
    for (let i = 0; i < newBBox.length; i++) bbox[i] = newBBox[i];
  }
  return newPoint;
}

File: rust/geometry/src/wm/convert/vector.rs
use alloc::{vec, vec::Vec};
use s2json::{
    BBox3D, BaseGeometry, Geometry, LineStringMValues, MultiLineStringMValues, MultiPolygonMValues,
    PointOrPoint3D, VectorGeometry, VectorPoint,
};

// TODO: Consider taking m-value data rather then clones

/// Convert a GeoJSON Geometry to an Vector Geometry
pub fn convert_geometry_to_vector<M: Clone + Default>(
    geometry: &Geometry<M>,
    build_bbox: bool,
) -> VectorGeometry<M> {
    let mut bbox = if build_bbox { Some(BBox3D::default()) } else { None };
    match geometry {
        Geometry::Point(geo) => {
            let coords = to_vector_point(geo, &mut bbox);
            VectorGeometry::new_point(coords, bbox)
        }
        Geometry::Point3D(geo) => {
            let coords = to_vector_point(geo, &mut bbox);
            VectorGeometry::new_point(coords, bbox)
        }
        Geometry::MultiPoint(geo) => {
            let coords = to_vector_multipoint(geo, &mut bbox);
            VectorGeometry::new_multipoint(coords, bbox)
        }
        Geometry::MultiPoint3D(geo) => {
            let coords = to_vector_multipoint(geo, &mut bbox);
            VectorGeometry::new_multipoint(coords, bbox)
        }
        Geometry::LineString(geo) => {
            let coords = to_vector_multipoint(geo, &mut bbox);
            VectorGeometry::new_linestring(coords, bbox)
        }
        Geometry::LineString3D(geo) => {
            let coords = to_vector_multipoint(geo, &mut bbox);
            VectorGeometry::new_linestring(coords, bbox)
        }
        Geometry::MultiLineString(geo) => {
            let coords = to_vector_multilinestring(geo, &mut bbox);
            VectorGeometry::new_multilinestring(coords, bbox)
        }
        Geometry::MultiLineString3D(geo) => {
            let coords = to_vector_multilinestring(geo, &mut bbox);
            VectorGeometry::new_multilinestring(coords, bbox)
        }
        Geometry::Polygon(geo) => {
            let coords = to_vector_multilinestring(geo, &mut bbox);
            VectorGeometry::new_polygon(coords, bbox)
        }
        Geometry::Polygon3D(geo) => {
            let coords = to_vector_multilinestring(geo, &mut bbox);
            VectorGeometry::new_polygon(coords, bbox)
        }
        Geometry::MultiPolygon(geo) => {
            let coords = to_vector_multipolygon(geo, &mut bbox);
            VectorGeometry::new_multipolygon(coords, bbox)
        }
        Geometry::MultiPolygon3D(geo) => {
            let coords = to_vector_multipolygon(geo, &mut bbox);
            VectorGeometry::new_multipolygon(coords, bbox)
        }
    }
}

fn convert_point<M: Clone + Default>(
    point: PointOrPoint3D,
    m: Option<M>,
    bbox: &mut Option<BBox3D>,
) -> VectorPoint<M> {
    let new_point = VectorPoint::new(point.0, point.1, point.2, m);
    if let Some(b) = bbox {
        b.extend_from_point(&new_point);
    }
    new_point
}

fn to_vector_point<M: Clone + Default, G: Copy, B>(
    geo: &BaseGeometry<M, G, B>,
    bbox: &mut Option<BBox3D>,
) -> VectorPoint<M>
where
    PointOrPoint3D: core::convert::From<G>,
{
    convert_point(geo.coordinates.into(), geo.m_values.clone(), bbox)
}

fn to_vector_multipoint<M: Clone + Default, G: Copy, B>(
    geo: &BaseGeometry<LineStringMValues<M>, Vec<G>, B>,
    bbox: &mut Option<BBox3D>,
) -> Vec<VectorPoint<M>>
where
    PointOrPoint3D: core::convert::From<G>,
{
    let binding = vec![];
    let m_values = geo.m_values.as_ref().unwrap_or(&binding);
    geo.coordinates
        .iter()
        .enumerate()
        .map(|(i, p)| convert_point((*p).into(), m_values.get(i).cloned(), bbox))
        .collect()
}

fn to_vector_multilinestring<M: Clone + Default, G: Copy, B>(
    geo: &BaseGeometry<MultiLineStringMValues<M>, Vec<Vec<G>>, B>,
    bbox: &mut Option<BBox3D>,
) -> Vec<Vec<VectorPoint<M>>>
where
    PointOrPoint3D: core::convert::From<G>,
{
    let binding = vec![];
    let m_values = geo.m_values.as_ref().unwrap_or(&binding);
    geo.coordinates
        .iter()
        .enumerate()
        .map(|(i, line)| {
            line.iter()
                .enumerate()
                .map(|(j, p)| {
                    let m_value = m_values.get(i).and_then(|m| m.get(j)).cloned();
                    convert_point((*p).into(), m_value, bbox)
                })
                .collect()
        })
        .collect()
}

fn to_vector_multipolygon<M: Clone + Default, G: Copy, B>(
    geo: &BaseGeometry<MultiPolygonMValues<M>, Vec<Vec<Vec<G>>>, B>,
    bbox: &mut Option<BBox3D>,
) -> Vec<Vec<Vec<VectorPoint<M>>>>
where
    PointOrPoint3D: core::convert::From<G>,
{
    let binding = vec![];
    let m_values = geo.m_values.as_ref().unwrap_or(&binding);
    geo.coordinates
        .iter()
        .enumerate()
        .map(|(i, polygon)| {
            polygon
                .iter()
                .enumerate()
                .map(|(j, line)| {
                    line.iter()
                        .enumerate()
                        .map(|(k, p)| {
                            let m_value = m_values
                                .get(i)
                                .and_then(|m| m.get(j))
                                .and_then(|m| m.get(k))
                                .cloned();
                            convert_point((*p).into(), m_value, bbox)
                        })
                        .collect()
                })
                .collect()
        })
        .collect()
}
